### PR TITLE
java 11 to 17

### LINKED ISSUE #11011: Update build image (created by /prime-router/build.sh) to Java 17 and other good things
The base image we use to build our application needs upgrades to:
Gradle - 
Azure Functions - looks like we are on the latest https://github.com/Azure/azure-functions-core-tools/releases
Java - 17

### PR DESCRIPTION
**Pull request by JosiahSiegel**
This PR updates build environments to use Java 17, as well as replaces the usage of an outdated docker image with a local build file.

Test Steps:
1. Run a docker hub build

## Changes
- updates build environments to use Java 17
- replaces the usage of an outdated docker image with a local build file
- Consolidate workflows to fetch from same java build action

## Linked Issues
- Fixes #11011

### CODE REVIEW

**Review by MauriceReeves-usds — COMMENTED** [review]
@JosiahSiegel I noticed in the `build.gradle.kts` that the Kotlin versions are not aligned. One spot says `1.8.22` and another says `1.9.10`. Those should probably be aligned.

**Comment by jack-h-wang** [comment]
@JosiahSiegel I would suggest building this off the work done in #10992

File: prime-router/build.gradle.kts [hunk 1]
```diff
@@ -34,7 +34,7 @@ plugins {
     id("org.flywaydb.flyway") version "9.21.2"
     id("nu.studer.jooq") version "7.1.1"
     id("com.github.johnrengelman.shadow") version "7.1.2"
-    id("com.microsoft.azure.azurefunctions") version "1.13.0"
```
Comment (jack-h-wang) [inline prime-router/build.gradle.kts:37]: This seems to be reverting a recently merged dependabot update. We should probably leave this alone.

File: .github/actions/build-backend/action.yml [hunk 2]
```diff
@@ -46,9 +76,121 @@ runs:
...
+        rm -f .gradle/caches/modules-2/gc.properties
+      shell: bash
+
     - name: Upload Artifact
       uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32
       with:
         name: prime-router-build-${{ inputs.version }}
         path: prime-router/prime-router-build.tar.gz
         retention-days: 3
+
```
Comment (snesm) [inline .github/actions/build-backend/action.yml:169]: These steps were previously a separate job.  They are likely to add 10 min. to the full build time.

File: .github/workflows/build_hub.yml [hunk 3]
```diff
@@ -190,55 +102,16 @@ jobs:
...
-        uses: github/codeql-action/analyze@4fddc51e4f3b5e5e9022f35c3464736cc10e1e98
-        if: github.actor != 'dependabot[bot]'
-
-      - name: "Run SonarCloud Scan"
-        uses: sonarsource/sonarcloud-github-action@db501078e936e4b4c8773d1bb949ba9ddb7b6b6a
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
-        if: github.actor != 'dependabot[bot]'
+      - name: Build backend
```
Comment (snesm) [inline .github/workflows/build_hub.yml:110]: The entire docker_build_test job seems unnecessary since you've moved all the steps to build-backend.

**Review by snesm — CHANGES_REQUESTED** [review]

**Comment by snesm** [comment]
FYI: jvm updates in [progress](https://github.com/CDCgov/prime-reportstream/pull/11227)

**Comment by snesm** [comment]
> > FYI: jvm updates in [progress](https://github.com/CDCgov/prime-reportstream/pull/11227)
> 
> @jack-h-wang @MauriceReeves-usds @snesm @luis-pabon-tf . Did you want to wait on the JVM update above before merging Java 17?

I believe the kotlin (related) JVM is independent except this PR unifies on 1.8.22, and https://github.com/CDCgov/prime-reportstream/pull/11227 unifies on 1.9.10.

Either is probably better than the current mixed versions in master.

**Review by snesm — APPROVED** [review]

File: .github/actions/build-backend/action.yml [hunk 4]
```diff
@@ -46,9 +76,121 @@ runs:
...
+    - name: Perform Java CodeQL Analysis
+      if: github.actor != 'dependabot[bot]' && inputs.run-qc == 'true'
+      uses: github/codeql-action/analyze@4fddc51e4f3b5e5e9022f35c3464736cc10e1e98
+
+    - name: Run SonarCloud Scan
+      if: |
+        github.actor != 'dependabot[bot]' && inputs.run-qc == 'true' &&
+        inputs.github-token != 'false' && inputs.sonar-token != 'false'
+      uses: sonarsource/sonarcloud-github-action@db501078e936e4b4c8773d1bb949ba9ddb7b6b6a
```
Comment (jack-h-wang) [inline .github/actions/build-backend/action.yml:178]:
```suggestion
      uses: sonarsource/sonarcloud-github-action@c25d2e7e3def96d0d1781000d3c429da22cd6252
```

File: .github/actions/build-frontend/action.yml [hunk 5]
```diff
@@ -92,7 +92,7 @@ runs:
 
     - name: SonarCloud Scan
       if: github.actor != 'dependabot[bot]' && inputs.GITHUB_TOKEN != '' && inputs.SONAR_TOKEN != ''
-      uses: sonarsource/sonarcloud-github-action@db501078e936e4b4c8773d1bb949ba9ddb7b6b6a
+      uses: sonarsource/sonarcloud-github-action@4b4d7634dab97dcee0b75763a54a6dc92a9e6bc1
```
Comment (jack-h-wang) [inline .github/actions/build-frontend/action.yml:95]:
```suggestion
      uses: sonarsource/sonarcloud-github-action@c25d2e7e3def96d0d1781000d3c429da22cd6252
```

File: .github/workflows/sonarcloud_master.yml [hunk 6]
```diff
@@ -54,8 +41,8 @@ jobs:
         run: yarn jest --coverage --silent --maxWorkers=100%
         working-directory: frontend-react
 
-      - name: "Run SonarCloud Scan"
-        uses: sonarsource/sonarcloud-github-action@db501078e936e4b4c8773d1bb949ba9ddb7b6b6a
+      - name: Run SonarCloud Scan
+        uses: sonarsource/sonarcloud-github-action@4b4d7634dab97dcee0b75763a54a6dc92a9e6bc1
```
Comment (jack-h-wang) [inline .github/workflows/sonarcloud_master.yml:45]:
```suggestion
        uses: sonarsource/sonarcloud-github-action@c25d2e7e3def96d0d1781000d3c429da22cd6252
```

**Review by jack-h-wang — CHANGES_REQUESTED** [review]
One of the SonarCloud action hashes was still set to the older Java 11 release. Not sure if there's an easier way of unifying all the instances of SonarCloud scanning behind a single action hash, but I've highlighted all the ones I could find and filled in the latest one.

**Review by jack-h-wang — APPROVED** [review]

File: .github/actions/sonarcloud/action.yml
Comment (snesm) [inline .github/actions/sonarcloud/action.yml:1]: SonarCloud scans are fragile.  The key seems to be providing the required build outputs.  Running the scans without the required build will **delete** SonarCloud data.

**Pull request merged**